File: Ball.cpp
#include "Ball.h"

#ifdef __APPLE__
#include <GLUT/glut.h>
#else
#include <GL/glut.h>
#endif

double damping = 0.95;
//double collisionDamping = 0.3;
double collisionDamping = 0.3;

Ball::Ball() {
  center.SetAll(0,0,0);
  velocity.SetAll(0,0,0);
  accel.SetAll(0,0,0);
  color.SetAll(0.3,0.2,0.8);
}

void Ball::SetValues(double _radius, Vector3d _center, Vector3d _vel, Vector3d _accel, double* _bbx) {
  radius = _radius;
  center = _center;
  velocity = _vel;
  accel = _accel;
  bbx = _bbx;
}

void Ball::SetRandomColor() {
  color.SetX( drand48() );
  color.SetY( drand48() );
  color.SetZ( drand48() );
}

bool Ball::IsMoving() {
  if( velocity.norm() < 0.2 ) return true;
  else return false;
}

void Ball::Update(double dt) {
  //cout << "Ball update! center = " << center << " vel = " << velocity << " accel = " << accel << endl;
  center = center + velocity * dt;
  oldVelocity = velocity;
  velocity = velocity + accel * dt;
  double oldAccelY = accel.GetY();
  oldAccelY -= 9.8;
  accel.SetY(oldAccelY);
  accel *= damping; 

  ResolveCollision(); //this should fix things in the case that it goes out of bounds

  if(accel.norm() < 0.5 ) accel.SetAll(0,0,0);
  if(velocity.norm() < 0.5 ) velocity.SetAll(0,0,0);
}

//              [  0 , 1  ,  2  , 3  ,  4  , 5  ]
// bbx contains [xMin,xMax, yMin,yMax, zMin,zMax]
void Ball::ResolveCollision() {
  //resolve X component
  if( center.GetX()-radius < bbx[0] ) {
    double oldCenterX = center.GetX();
    //make sure center doesn't cross plane
    center.SetX(bbx[0]+radius);
    //reflect velocity - just change y direction
    velocity.SetX( -1*velocity.GetX() );
    velocity *= collisionDamping;
    accel *= collisionDamping;
  }
  if( center.GetX()+radius > bbx[1] ) {
    double oldCenterX = center.GetX();
    //make sure center doesn't cross plane
    center.SetX(bbx[1]-radius);
    //reflect velocity - just change y direction
    velocity.SetX( -1*velocity.GetX() );
    velocity *= collisionDamping;
    accel *= collisionDamping;
  }
  //resolve Y component
  if( center.GetY()-radius < bbx[2] ) {
    double oldCenterY = center.GetY();
    //make sure center doesn't cross plane
    center.SetY(bbx[2]+radius);
    //reflect velocity - just change y direction
    velocity.SetY( -1*velocity.GetY() );
    //accel.SetY( -1*accel.GetY() );
    velocity *= collisionDamping;
    accel *= collisionDamping;
  }
  if( center.GetY()+radius > bbx[3] ) {
    double oldCenterY = center.GetY();
    //make sure center doesn't cross plane
    center.SetY(bbx[3]-radius);
    //reflect velocity - just change y direction
    velocity.SetY( -1*velocity.GetY() );
    velocity *= collisionDamping;
    accel *= collisionDamping;
  }
  //resolve Z component
  if( center.GetZ()-radius < bbx[4] ) {
    double oldCenterZ = center.GetZ();
    //make sure center doesn't cross plane
    center.SetZ(bbx[4]+radius);
    //reflect velocity - just change y direction
    velocity.SetZ( -1*velocity.GetZ() );
    velocity *= collisionDamping;
    accel *= collisionDamping;
  }
  if( center.GetZ()+radius > bbx[5] ) {
    double oldCenterZ = center.GetZ();
    //make sure center doesn't cross plane
    center.SetZ(bbx[5]-radius);
    //reflect velocity - just change y direction
    velocity.SetZ( -1*velocity.GetZ() );
    velocity *= collisionDamping;
    accel *= collisionDamping;
  }
}

void Ball::Draw() {
  //cout << "Ball draw!" << endl;

  //glColor3f(0.3,0.2,0.8);
  glColor3f(color.GetX(),color.GetY(),color.GetZ());
  glPushMatrix();
  glTranslated(center[0],center[1],center[2]);
  glutSolidSphere(radius,10,10);
  glPopMatrix();

  /*
  glPointSize(3);
  glColor3f(1,0,0);
  glBegin(GL_POINTS);
  glVertex3d(center[0],center[1],center[2]);
  glEnd();
  */
}

File: py/Ball.py
import Vector
import random
import math
from OpenGL.GLUT import *
from OpenGL.GLU import *
from OpenGL.GL import *

damping = 0.95
collisionDumping = 0.3



class Ball:
    def __init__(self): 
        self.center = Vector.Vector3D(0, 0, 0)
        self.velocity = Vector.Vector3D(0, 0, 0)
        self.accel = Vector.Vector3D(0, 0, 0)
        self.color = Vector.Vector3D(1, 1, 1)
        self.bbx = None
        self.radius = 0
        # self.ball = None
        # self.next = None

    def print_ball(self):
        self.center.print_vec()
        self.velocity.print_vec()
        self.accel.print_vec()
        self.color.print_vec()
        print(self.bbx)

    def SetValues(self, _radius, _center, _vel, _accel, _bbx):
        self.center = _center
        self.radius = _radius
        self.velocity = _vel
        self.accel = _accel
        self.bbx = _bbx

    def SetRandomColor(self):
        self.color.SetX(random.random())
        self.color.SetY(random.random())
        self.color.SetZ(random.random())

    def IsMoving(self):
        v = self.velocity.norm()
        if v < 0.2:
            return True
        else:
            return False

    def Update(self, dt, blist):
        self.center += self.velocity * dt
        oldVelocity = self.velocity
        self.velocity += self.accel * dt
        oldAccelY = self.accel.GetY()
        oldAccelY -= 9.8
        self.accel.SetY(oldAccelY)
        self.accel *= damping

        self.ResolveCollision()
        for i in blist:
            self.resolve_ball_ball(i)

        if self.accel.norm() < 0.5:
            self.accel.Set_All(0, 0, 0)
        if self.velocity.norm() < 0.5:
            self.velocity.Set_All(0, 0, 0)
        # print(type(rt_val))
        # print(rt_val[0])
        # return myt
        # return rt_val
        # return my_t

    def Draw(self):
        glColor3f(self.color.GetX(), self.color.GetY(), self.color.GetZ())
        glPushMatrix()
        glTranslated(self.center.GetX(), self.center.GetY(), self.center.GetZ())
        glutSolidSphere(self.radius, 100, 100)
        glPopMatrix()

    def ResolveCollision(self):
        if self.center.GetX() - self.radius < self.bbx[0]:
            oldCenterX = self.center.GetX()
            self.center.SetX(self.bbx[0] + self.radius)
            self.velocity.SetX(-1 * self.velocity.GetX() )
            self.velocity *= collisionDumping
            self.accel *= collisionDumping

        if self.center.GetX() + self.radius > self.bbx[1]:
            oldCenterX = self.center.GetX()
            self.center.SetX(self.bbx[1] - self.radius)
            self.velocity.SetX(-1 * self.velocity.GetX() )
            self.velocity *= collisionDumping
            self.accel *= collisionDumping

        if self.center.GetY() - self.radius <= self.bbx[2]:
            oldCenterX = self.center.GetY()
            self.center.SetY(self.bbx[2] + self.radius)
            self.velocity.SetY(-1 * self.velocity.GetY() )
            # self.velocity *= -1
            self.velocity *= collisionDumping
            self.accel *= collisionDumping
            # if self.accel.norm() < 0.5:
            #     self.accel.Set_All(0, 0, 0)
            # if self.velocity.norm() < 0.1:
            #     self.velocity.Set_All(0, 0, 0)
            # if self.accel.GetY() < 0.5:
            #     self.accel.SetY(0)

        if self.center.GetY() + self.radius > self.bbx[3]:
            oldCenterX = self.center.GetY()
            self.center.SetY(self.bbx[3] - self.radius)
            self.velocity.SetY(-1 * self.velocity.GetY() )
            self.velocity *= collisionDumping
            self.accel *= collisionDumping
            

        if self.center.GetZ() - self.radius < self.bbx[4]:
            oldCenterX = self.center.GetZ()
            self.center.SetZ(self.bbx[4] + self.radius)
            self.velocity.SetZ(-1 * self.velocity.GetZ() )
            self.velocity *= collisionDumping
            self.accel *= collisionDumping

        if self.center.GetZ() + self.radius > self.bbx[5]:
            oldCenterX = self.center.GetZ()
            self.center.SetZ(self.bbx[5] - self.radius)
            self.velocity.SetZ(-1 * self.velocity.GetZ() )
            self.velocity *= collisionDumping
            self.accel *= collisionDumping

    def get_center(self):
        return self.center
    
    # def resolve_ball_ball(self, ball_List : list):
    #     for i in range(len(ball_List)):
    #         for j in range(len(ball_List)):
    #             if self.detect_collision(ball_two) == True and i != j :
                    

    def resolve_ball_ball(self, _another):
        if self.detect_collision(_another) and self != _another: 
            _vel_one = self.velocity
            # _vel_one.SetY(0)
            _vel_two = _another.velocity
            # _vel_two.SetY(0)
            _accel_one = self.accel
            # _accel_one.SetY(0)
            _accel_two = _another.accel
            # _accel_two.SetY(0)
            # _vel_sum = _vel_one + _vel_two
            
            # _another.velocity = _vel_one
            # _another.accel = _accel_one
            # self.velocity = _vel_two
            # self.accel = _accel_two
            # print("inside")
            
            _another.velocity = (_vel_one + _vel_two)/2
            # _another.accel = _accel_one
            self.velocity = -(_vel_one + _vel_two)/2
            # self.accel = _accel_two



    def detect_collision(self, another_ball):
        other_ball_radius = another_ball.center
        if abs((another_ball.get_center() - self.get_center()).norm()) - (self.radius * 2) <= 0.0001:
            # if self != another_ball:
            return True
        else:
            return False



class node:
    def __init__(self):
        v = Vector.Vector3D(0, 0, 0)
        b = Ball()
        self.ball = b
        self.next = None

    def getBall(self):
        return self.ball
    
    def setBall(self, a):
        self.ball = a
